src/pumps/batch.rs

```diff
@@ -0,0 +1,73 @@
+use tokio::{
+    sync::mpsc::{self, Receiver},
+    task::JoinHandle,
+};
+
+use crate::Pump;
+
+pub struct BatchPump<T> {
+    pub(crate) n: usize,
+    batch: Vec<T>,
+}
+
+impl<T> BatchPump<T> {
+    pub fn new(n: usize) -> Self {
+        Self {
+            n,
+            batch: Vec::with_capacity(n),
+        }
+    }
+}
+
+impl<T> Pump<T, Vec<T>> for BatchPump<T>
+where
+    T: Send + 'static,
+{
+    fn spawn(mut self, mut input_receiver: Receiver<T>) -> (Receiver<Vec<T>>, JoinHandle<()>) {
+        let (output_sender, output_receiver) = mpsc::channel(1);
+
+        let h = tokio::spawn(async move {
+            while let Some(input) = input_receiver.recv().await {
+                self.batch.push(input);
+                if self.batch.len() >= self.n {
+                    let batch = std::mem::replace(&mut self.batch, Vec::with_capacity(self.n));
+                    if let Err(_e) = output_sender.send(batch).await {
+                        break;
+                    }
+                }
+            }
+            if !self.batch.is_empty() {
+                let _ = output_sender.send(self.batch).await;
```

Comment:
add a comment explaining why we don't care about the result of the send - if the output channel was dropped - there's nothing we can do here